File: solutions/cpp/beer-song/1/beer_song.cpp
#include "beer_song.h"

#include <string>
using namespace std;

namespace beer_song {

string container(int n, bool capital=false){
    if (!n)
       return  capital ?   "No more bottles " :  "no more bottles ";
    else
        return to_string(n) + ((n>1) ? " bottles " : " bottle ");
}

string verse(int n){
    string line1, line2;
    line1 = container(n,true) + "of beer on the wall, " + container(n,false) +
        "of beer.\n";
    if(n) {          
        string one_of_em = (n>1 ? "one" : "it");
        line2 = "Take " + one_of_em + " down and pass it around, " +                         container(n-1) + "of beer on the wall.\n";
    } else {
          line2 = "Go to the store and buy some more, " + container(99) + 
              "of beer on the wall.\n";
    }
    return line1 + line2;    
}

string sing(int from, int to){
    string song;
    for (int i = from; i >= to; i--){
        song = song + verse(i);
        if(i>to) 
           song.push_back('\n');
    }  
    return song;
}

}  // namespace beer_song

File: solutions/clojure/beer-song/1/src/beer_song.clj
(ns beer-song)

(defn bottles [num]
  (if (= num 1) " bottle "
    " bottles ")
  )

(defn verse
  "Returns the nth verse of the song."
  [num]
  (let [dnum (dec num)]
    (if (>= num 1) 
      (str num (bottles num) "of beer on the wall, " num (bottles num) "of beer.\n"
         "Take "  (if (> num 1) "one" "it") " down and pass it around, " (if (zero? dnum) "no more" (str dnum)) (bottles dnum) "of beer on the wall.\n" ) 
         (str "No more bottles of beer on the wall, no more bottles of beer.\n" "Go to the store and buy some more, 99 bottles of beer on the wall.\n")
      )))

(defn sing
  "Given a start and an optional end, returns all verses in this interval. If
  end is not given, the whole song from start is sung."
  ([start] (sing start 0))
  ([start end]
   (clojure.string/join "\n" (map verse (range start (dec end) -1)))
 ))  
  
